bundlewrap/operations.py

```diff
@@ -249,6 +256,52 @@ def run(
     return result
 
 
+def run_routeros(hostname, username, password, *args):
+    with ROUTEROS_CONNECTIONS_LOCK:
+        try:
+            conn_state = ROUTEROS_CONNECTIONS[hostname]
+        except KeyError:
+            conn_state = {
+                'connection': None,
+                'lock': Lock(),
+                'needs_reconnect': True,
+            }
+            ROUTEROS_CONNECTIONS[hostname] = conn_state
+
+    with conn_state['lock']:
+        if conn_state['needs_reconnect']:
+            if conn_state['connection']:
+                with suppress(Exception):
+                    conn_state['connection'].close()
+
+            try:
+                conn_state['connection'] = connect(
+                    # str() to resolve Faults
+                    username=str(username),
+                    password=str(password or ""),
+                    host=hostname,
+                    timeout=120.0,
+                )
+            except Exception as e:
+                raise RemoteException(str(e)) from e
+            else:
+                conn_state['needs_reconnect'] = False
+
+        try:
+            io.debug(f'{hostname}: running routeros command: {repr(args)}')
+            result = tuple(conn_state['connection'].rawCmd(*args))
```

Comment:
Code looks good to me. (In theory, `.close()` failing for some odd reason without actually closing the socket could exhaust the process’s file descriptor limit … Should we at least add an `io.debug()` there? The object we’re dealing with here is of type `librouteros.api.Api`, it’s not the raw socket, so errors in the library *are* a possibility.)

Can’t think of any other scenarios that break anything. So, added after line 292 to simulate errors:
```
            if randint(0, 10) > 8:
                raise Exception("Breaking on purpose")
```
Nothing unexpected showed up. Doesn’t mean *for certain* that there aren’t any errors, of course, but I have a good feeling about this.

I’m for merging this. Maybe add the `io.debug()`, your call.